### PR TITLE
coap-gateway: check and unpublish resources if the device doesn't contain them

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: coap-gateway/service/observation/deviceObserver.go [hunk 1]
```diff
@@ -327,6 +375,39 @@ func getPublishedResources(ctx context.Context, rdClient GrpcGatewayClient, devi
...
+	if len(publishedResources) == 0 {
+		return nil, nil
+	}
+	if len(deviceResources) == 0 {
+		return publishedResources, nil
+	}
+	validResources := make([]*commands.Resource, 0, len(publishedResources))
+	toUnpublishResourceInstanceIds := make([]int64, 0, len(publishedResources))
+	for _, res := range publishedResources {
+		if _, ok := deviceResources.GetResourceLink(res.GetHref()); ok {
```
Comment (Danielius1922) [inline coap-gateway/service/observation/deviceObserver.go:411]: Nitpick: This is O(number of published_resources x number of deviceResources) ~ O(n^2). With ordered arrays / sets we could get it down to O(n)

File: coap-gateway/service/observation/observation_test.go [hunk 2]
```diff
@@ -103,87 +97,22 @@ func TestIsResourceObservableWithInterface(t *testing.T) {
...
-		},
-		{
-			name: testResourceName(device.ResourceURI) + "with not supported observeInterface",
-			args: args{
-				resourceHref:     device.ResourceURI,
-				resourceType:     device.ResourceType,
-				observeInterface: interfaces.OC_IF_B,
-			},
-			want: false,
+			name:    testResourceName("/oic/res"),
```
Comment (Danielius1922) [inline coap-gateway/service/observation/observation_test.go]: "/oic/res" -> resources.ResourceURI

I would also add at least one case with `wantErr` true (invalid href maybe?) and `want` true (some obsevable resource).

File: coap-gateway/service/observation/deviceObserver_test.go [hunk 3]
```diff
@@ -259,8 +414,8 @@ func runTestDeviceObserverRegister(ctx context.Context, t *testing.T, deviceID s
 	// give time to wait for data
 	ctxWait, waitCancel := context.WithTimeout(context.Background(), time.Second*10)
 	closeWaitChans := func() {
-		close(retrieveChan)
-		close(observeChan)
+		//	close(retrieveChan)
```
Comment (Danielius1922) [inline coap-gateway/service/observation/deviceObserver_test.go]: Revert?

File: coap-gateway/service/observation/deviceObserver.go [hunk 4]
```diff
@@ -178,12 +216,27 @@ func emptyDeviceIDError() error {
...
+	}
+	resourceHref := resources.ResourceURI
+	observeInterface := interfaces.OC_IF_B
+	res, ok := links.GetResourceLink(resourceHref)
+	if !ok {
+		return false, fmt.Errorf("resourceLink for href(%v) not found", resourceHref)
+	}
+
+	observable := res.Policy.BitMask.Has(schema.Observable)
+	if observeInterface == "" || !observable {
```
Comment (Danielius1922) [inline coap-gateway/service/observation/deviceObserver.go]: `observeInterface == ""` is never true now

**Review by Danielius1922 — CHANGES_REQUESTED** [review]

File: coap-gateway/service/observation/deviceObserver.go (hunk 1 above)
Comment (jkralik) [reply coap-gateway/service/observation/deviceObserver.go:411]: fixed

File: coap-gateway/service/observation/observation_test.go (hunk 2 above)
Comment (jkralik) [reply coap-gateway/service/observation/observation_test.go]: done

File: coap-gateway/service/observation/deviceObserver_test.go (hunk 3 above)
Comment (jkralik) [reply coap-gateway/service/observation/deviceObserver_test.go]: fixed

File: coap-gateway/service/observation/deviceObserver.go (hunk 4 above)
Comment (jkralik) [reply coap-gateway/service/observation/deviceObserver.go]: fixed

File: coap-gateway/service/observation/deviceObserver.go [hunk 5]
```diff
@@ -375,15 +375,35 @@ func getPublishedResources(ctx context.Context, rdClient GrpcGatewayClient, devi
...
+	validResources = make([]*commands.Resource, 0, len(publishedResources))
+	toUnpublishResourceInstanceIds = make([]int64, 0, len(publishedResources))
+	publishedResources.Sort()
+	deviceResources.Sort()
+	var j int
+	for _, res := range publishedResources {
+		for j < len(deviceResources) && deviceResources[j].Href < res.GetHref() {
+			j++
+		}
+		if deviceResources[j].Href == res.GetHref() {
```
Comment (Danielius1922) [inline coap-gateway/service/observation/deviceObserver.go:391]: In the case that `publishedResources` has a resource with >Href than the last element of `deviceResources` then `j=len(deviceResources)`, which means `deviceResources[j].Href` will be out-of-bounds. 

The simple fix is to add
```
if j>=len(deviceResources) {
  break
}
```
after the inner loop.

File: coap-gateway/service/observation/observation.go [hunk 6]
```diff
@@ -6,15 +6,14 @@ import (
 
 	"github.com/plgd-dev/device/schema"
 	"github.com/plgd-dev/device/schema/interfaces"
-	"github.com/plgd-dev/device/schema/resources"
 	coapMessage "github.com/plgd-dev/go-coap/v2/message"
 	"github.com/plgd-dev/go-coap/v2/message/codes"
 	"github.com/plgd-dev/kit/v2/codec/cbor"
 )
 
 // Query /oic/res resource to get device resource links.
```
Comment (Danielius1922) [inline coap-gateway/service/observation/observation.go]: Docu no longer correct, other resources that support `if.ll` interface can now be queried too.

File: coap-gateway/service/observation/deviceObserver.go (hunk 5 above)
Comment (jkralik) [reply coap-gateway/service/observation/deviceObserver.go:391]: fixed

File: coap-gateway/service/observation/observation.go (hunk 6 above)
Comment (jkralik) [reply coap-gateway/service/observation/observation.go]: fixed

**Review by Danielius1922 — APPROVED** [review]

**Pull request merged**